### ISSUE TITLE
False positive on jar scanning on fixed older minor version than latest

### ISSUE THREAD

**Issue by mbreevoort** [aquasecurity/trivy#1025]
## Description

The nimbus-jose-jwt version 9.8.1  got shaded json-smart which got a vulnerability.
```
{
    "Target": "app/lib/nimbus-jose-jwt-9.8.1.jar",
    "Type": "jar",
    "Vulnerabilities": [
      {
        "VulnerabilityID": "CVE-2021-27568",
        "PkgName": "net.minidev:json-smart",
        "InstalledVersion": "1.3.2",
        "FixedVersion": "2.4.1",
        "Layer": {
          "DiffID": "sha256:c16fcd5fe87ea387a5899b20fc0324af4863342b0ee962bce0cbda7464cb9f19"
        },
        "SeveritySource": "nvd",
        "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2021-27568",
        "Title": "json-smart: uncaught exception may lead to crash or information disclosure",
        "Description": "An issue was discovered in netplex json-smart-v1 through 2015-10-23 and json-smart-v2 through 2.4. An exception is thrown from a function, but it is not caught, as demonstrated by NumberFormatException. When it is not caught, it may cause programs using the library to crash or expose sensitive information.",
        "Severity": "CRITICAL",
        "CweIDs": [
          "CWE-754"
        ],
```

When I go to https://nvd.nist.gov/vuln/detail/CVE-2021-27568 :
`Up to (excluding) 1.3.2` -> the shaded version
`Up to (excluding) 2.3.1`
`Up to (excluding) 2.4.1`

## What did you expect to happen?
No vulnerability on an excluded not latest version

## What happened instead?
"Severity": "CRITICAL" on not vulnerable version 1.3.2

## Output of run with `-debug`:

```
{
    "Target": "app/lib/nimbus-jose-jwt-9.8.1.jar",
    "Type": "jar",
    "Vulnerabilities": [
      {
        "VulnerabilityID": "CVE-2021-27568",
        "PkgName": "net.minidev:json-smart",
        "InstalledVersion": "1.3.2",
        "FixedVersion": "2.4.1",
        "Layer": {
          "DiffID": "sha256:c16fcd5fe87ea387a5899b20fc0324af4863342b0ee962bce0cbda7464cb9f19"
        },
        "SeveritySource": "nvd",
        "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2021-27568",
        "Title": "json-smart: uncaught exception may lead to crash or information disclosure",
        "Description": "An issue was discovered in netplex json-smart-v1 through 2015-10-23 and json-smart-v2 through 2.4. An exception is thrown from a function, but it is not caught, as demonstrated by NumberFormatException. When it is not caught, it may cause programs using the library to crash or expose sensitive information.",
        "Severity": "CRITICAL",
        "CweIDs": [
          "CWE-754"
        ],
```

## Additional details (base image name, container registry info...):
trivy 0.18.3

**Comment by masahiro331** [comment]
@mbreevoort 
Thank you report issue. 
This advisory is provided by the gitlab advisory database.

I will send a PR to fix the Adivosory.
https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/merge_requests/6889

**Comment by masahiro331** [comment]
PR merged. 
I think it will be reflected in Trivy DB the day after tomorrow.

**Comment by mbreevoort** [comment]
Thnx for fixing the gitlab report.
Tested and no false positive on nimbus-jose-jwt-9.8.1.jar -> shadowed json-smart 1.3.2 anymore :-)

**Comment by masahiro331** [comment]
If Fixed this probrem, please close issue!!

**Comment by mbreevoort** [comment]
Fixed

**Comment by mbreevoort** [comment]
I don't understand the problem is back but now for version 2.3.1
```
"Target": "app/lib/json-smart-2.3.1.jar",
    "Type": "jar",
    "Vulnerabilities": [
      {
        "VulnerabilityID": "CVE-2021-27568",
        "PkgName": "net.minidev:json-smart",
        "InstalledVersion": "2.3.1",
        "FixedVersion": "2.4.1, 1.3.2",
        "Layer": {
          "Digest": "sha256:d08152a7e00160722cbe6b8338d2a02049e02cc0e962cb7b139f6b4494636ca9",
          "DiffID": "sha256:9181a53001160e5d4a2b141afa04103990d2fe0e43b87f83564c9651e1f06107"
        },
        "SeveritySource": "nvd",
        "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2021-27568",
```

In https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/blob/master/maven/net.minidev/json-smart/CVE-2021-27568.yml version 2.3.1 is not vulnerable

**Comment by mbreevoort** [comment]
@masahiro331 can you help me how to fix the false positive for version 2.3.1 ?

**Comment by monwolf** [comment]
I'm getting the same behaviour using the version 2.4.7 
```
2021-07-16T10:30:16.312+0200    DEBUG   Parsing Java artifacts...       {"file": "BOOT-INF/lib/json-path-2.5.0.jar"}
2021-07-16T10:30:16.423+0200    DEBUG   Parsing Java artifacts...       {"file": "BOOT-INF/lib/json-smart-2.4.7.jar"}
2021-07-16T10:30:16.423+0200    DEBUG   Parsing Java artifacts...       {"file": "BOOT-INF/lib/accessors-smart-2.4.7.jar"}
```
And in the report we are getting:

```
+---------------------------------------------------+------------------+----------+-------------------+------------------------------+---------------------------------------+
|                      LIBRARY                      | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |        FIXED VERSION         |                 TITLE                 |
+---------------------------------------------------+------------------+----------+-------------------+------------------------------+---------------------------------------+
| log4j:log4j                                       | CVE-2019-17571   | CRITICAL | 1.2.8             |                              | log4j: deserialization of             |
|                                                   |                  |          |                   |                              | untrusted data in SocketServer        |
|                                                   |                  |          |                   |                              | -->avd.aquasec.com/nvd/cve-2019-17571 |
+---------------------------------------------------+------------------+          +-------------------+------------------------------+---------------------------------------+
| net.minidev:json-smart                            | CVE-2021-27568   |          | 1.3.2             | 2.3.1, 2.4.1                 | json-smart: uncaught                  |
|                                                   |                  |          |                   |                              | exception may lead to crash           |
|                                                   |                  |          |                   |                              | or information disclosure             |
|                                                   |                  |          |                   |                              | -->avd.aquasec.com/nvd/cve-2021-27568 |
+---------------------------------------------------+------------------+----------+-------------------+------------------------------+---------------------------------------+
| org.springframework.security:spring-security-core | CVE-2021-22119   | HIGH     | 5.5.0             | 5.2.11, 5.3.10, 5.4.7, 5.5.1 | spring-security: Denial-of-Service    |
|                                                   |                  |          |                   |                              | (DoS) attack via initiation           |
|                                                   |                  |          |                   |                              | of Authorization Request              |
|                                                   |                  |          |                   |                              | -->avd.aquasec.com/nvd/cve-2021-22119 |
+---------------------------------------------------+------------------+----------+-------------------+------------------------------+---------------------------------------+
```

**Comment by mbreevoort** [comment]
That's strange. I still get
```
+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
|        LIBRARY         | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                 TITLE                 |
+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
| net.minidev:json-smart | CVE-2021-27568   | CRITICAL | 2.3.1             | 2.4.1, 1.3.2  | json-smart: uncaught                  |
|                        |                  |          |                   |               | exception may lead to crash           |
|                        |                  |          |                   |               | or information disclosure             |
|                        |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2021-27568 |
+------------------------+------------------+----------+-------------------+---------------+---------------------------------------+
```
With fixed version 2.4.1 and 1.3.2
@monwolf is your trivy database updated?

**Comment by monwolf** [comment]
```
2021-07-16T10:29:49.394+0200    DEBUG   cache dir:  /var/lib/jenkins/trivy-scan/.trivycache/
2021-07-16T10:29:49.394+0200    DEBUG   DB update was skipped because DB is the latest
2021-07-16T10:29:49.394+0200    DEBUG   DB Schema: 1, Type: 1, UpdatedAt: 2021-07-16 06:04:12.248578955 +0000 UTC, NextUpdate: 2021-07-16 12:04:12.248578155 +0000 UTC, DownloadedAt: 2021-07-16 08:25:55.285537853 +0000 UTC
```
Today is the last time it was updated

**Comment by mbreevoort** [comment]
The gitlab bot committed wrong ranges again: https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/commit/bf86ae0c1236d9a0f4d26eba08fff3786be4b5f7

**Comment by mbreevoort** [comment]
Created an issue at gitlab https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/issues/165#note_644364393

**Comment by masahiro331** [comment]
@mbreevoort 
I'm sorry, I was away from work for a month due to illness.

Is there anything I can do to help?

**Comment by knqyf263** [comment]
Looks like it was fixed again in GitLab. I'll close this issue, but please feel free to re-open if needed. Thank you for reporting it!

**Comment by mbreevoort** [comment]
> @mbreevoort
> I'm sorry, I was away from work for a month due to illness.
> 
> Is there anything I can do to help?

No problem, hope you're well again. 
It is fixed in master from version v1.4.171
https://gitlab.com/gitlab-org/security-products/gemnasium-db/-/blob/v1.4.171/maven/net.minidev/json-smart/CVE-2021-27568.yml

The trivy scan still reports the error, how can I check how trivy connects to the gitlab gemnasium-db?
```
$ trivy --version   
Version: 0.19.2
Vulnerability DB:
  Type: Light
  Version: 1
  UpdatedAt: 2021-08-09 06:04:54.683685452 +0000 UTC
  NextUpdate: 2021-08-09 12:04:54.683685252 +0000 UTC
  DownloadedAt: 2021-08-09 09:10:22.069805 +0000 UTC
```

**Comment by manish7277** [comment]
I am facing these vul while checking trivy report 
+-------------+------------------+----------+-------------------+---------------+---------------------------------------+
| log4j:log4j | CVE-2019-17571   | CRITICAL | 1.2.17            |               | log4j: deserialization of             |
|             |                  |          |                   |               | untrusted data in SocketServer        |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2019-17571 |
+             +------------------+          +                   +---------------+---------------------------------------+
|             | CVE-2022-23305   |          |                   |               | log4j: SQL injection in               |
|             |                  |          |                   |               | Log4j 1.x when application            |
|             |                  |          |                   |               | is configured to use...               |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-23305 |
+             +------------------+          +                   +---------------+---------------------------------------+
|             | CVE-2022-23307   |          |                   |               | log4j: Unsafe deserialization         |
|             |                  |          |                   |               | flaw in Chainsaw log viewer           |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-23307 |
+             +------------------+----------+                   +---------------+---------------------------------------+
|             | CVE-2021-4104    | HIGH     |                   |               | log4j: Remote code execution          |
|             |                  |          |                   |               | in Log4j 1.x when application         |
|             |                  |          |                   |               | is configured to...                   |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2021-4104  |
+             +------------------+          +                   +---------------+---------------------------------------+
|             | CVE-2022-23302   |          |                   |               | log4j: Remote code execution          |
|             |                  |          |                   |               | in Log4j 1.x when application         |
|             |                  |          |                   |               | is configured to...                   |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-23302 |
+             +------------------+----------+                   +---------------+---------------------------------------+
|             | CVE-2020-9488    | LOW      |                   | 2.13.2        | log4j: improper validation            |
|             |                  |          |                   |               | of certificate with host              |
|             |                  |          |                   |               | mismatch in SMTP appender             |
|             |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2020-9488  |
+             +------------------+----------+                   +---------------+---------------------------------------+
|             | GMS-2021-5       | UNKNOWN  |                   | 2.15.0-rc1    | Improper Neutralization               |
|             |                  |          |                   |               | of Special Elements in                |
|             |                  |          |                   |               | Output Used by a Downstream           |
|             |                  |          |                   |               | Component...
